app/src/test/java/com/asfoundation/wallet/Gherkin.kt

```diff
@@ -0,0 +1,499 @@
+    callsInPlace(test, InvocationKind.EXACTLY_ONCE)
+  }
+  Gherkin.calls.clear()
+  Gherkin.names.clear()
+  Gherkin.test()
+  require(Gherkin.calls.contains(0)) { "'Given' is missing" }
+  require(Gherkin.calls.contains(1)) { "'When' is missing" }
+  require(Gherkin.calls.contains(2)) { "'Then' is missing" }
+}
+
+fun coGherkin(test: suspend Gherkin.(scope: TestScope) -> Unit) {
+  contract {
+    callsInPlace(test, InvocationKind.EXACTLY_ONCE)
+  }
+  Gherkin.calls.clear()
+  Gherkin.names.clear()
+  runTest {
+    Gherkin.test(this)
+  }
+  require(Gherkin.calls.contains(0)) { "'Given' is missing" }
+  require(Gherkin.calls.contains(1)) { "'When' is missing" }
+  require(Gherkin.calls.contains(2)) { "'Then' is missing" }
+}
+
+inline fun Gherkin.Given(desc: String, block: GherkinClass.() -> Unit): GherkinClass {
+  contract {
+    callsInPlace(block, InvocationKind.EXACTLY_ONCE)
+  }
+  require(desc.isNotBlank()) { "'Given' description should not be blank" }
+  require(calls.contains(0).not()) { "Only one 'Given' is allowed" }
+  require(calls.isEmpty()) { "'Given' should be the first one to call" }
+  require(names.contains(desc).not()) { "Duplicated description: $desc" }
+  GherkinClass.block()
+  calls.add(0)
+  names.add(desc)
+  return GherkinClass
+}
+
+inline fun Gherkin.When(desc: String, block: GherkinClass.() -> Unit): GherkinClass {
+  contract {
+    callsInPlace(block, InvocationKind.EXACTLY_ONCE)
+  }
+  require(desc.isNotBlank()) { "'When' description should not be blank" }
+  require(calls.contains(0)) { "'When' should be called after 'Given'" }
+  require(calls.contains(1).not()) { "Only one 'When' is allowed" }
+  require(names.contains(desc).not()) { "Duplicated description: $desc" }
+  GherkinClass.block()
+  calls.add(1)
+  names.add(desc)
+  return GherkinClass
+}
+
+inline fun Gherkin.Then(desc: String, block: GherkinClass.() -> Unit): GherkinClass {
+  contract {
+    callsInPlace(block, InvocationKind.EXACTLY_ONCE)
+  }
+  require(desc.isNotBlank()) { "'Then' description should not be blank" }
+  require(calls.contains(1)) { "'Then' should be called after 'When'" }
+  require(calls.contains(2).not()) { "Only one 'Then' is allowed" }
+  require(names.contains(desc).not()) { "Duplicated description: $desc" }
+  GherkinClass.block()
+  calls.add(2)
+  names.add(desc)
+  return GherkinClass
+}
+
+inline fun GherkinClass.and(desc: String, block: GherkinClass.() -> Unit): GherkinClass {
+  contract {
+    callsInPlace(block, InvocationKind.EXACTLY_ONCE)
+  }
+  require(Gherkin.names.contains(desc).not()) { "Duplicated description: $desc" }
+  block()
+  Gherkin.names.add(desc)
+  return this
+}
+
+inline fun GherkinClass.but(desc: String, block: GherkinClass.() -> Unit): GherkinClass {
+  contract {
+    callsInPlace(block, InvocationKind.EXACTLY_ONCE)
+  }
+  require(Gherkin.names.contains(desc).not()) { "Duplicated description: $desc" }
+  block()
+  Gherkin.names.add(desc)
+  return this
+}
+
+class GherkinTests {
```

Comment:
I would suggest moving the GherkinTests class into it's own file separate from the Gherkin.

Reply:
Done